MREUnityRuntime/MREUnityRuntimeLib/Factories/DefaultMaterialPatcher.cs

```diff
@@ -36,7 +36,7 @@ public void ApplyMaterialPatch(Material material, MWMaterial patch)
             }
             else if (patch.MainTextureId != null)
             {
-                material.mainTexture = (Texture)MREAPI.AppsAPI.AssetCache.GetAsset(patch.MainTextureId.Value);
+                MREAPI.AppsAPI.AssetCache.OnCached(patch.MainTextureId.Value, tex => material.mainTexture = (Texture)tex);
             }
```

Comment:
This fixes defers setting the value, which is good, there is a risk that other changes to material.mainTexture may come in out of order. For example: If there is a call to set texture to an uncached texture instantly followed by setting texture to null (or a cached texture), then the uncached value would be set afterwards.

Reply:
Agreed, this implementation leads to buggy behavior. Changes should be applied in order.

Reply:
Resolved by tracking what texture is supposed to be assigned, and returning early if things aren't right.

---
In reply to: [306960603](https://github.com/microsoft/mixed-reality-extension-unity/pull/92#discussion_r306960603) [](ancestors = 306960603)